test/jdk/java/net/httpclient/HttpResponseConnectionLabelTest.java

```diff
@@ -0,0 +1,444 @@
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.ForkJoinPool;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.stream.Stream;
+
+import static java.net.http.HttpClient.Builder.NO_PROXY;
+import static java.nio.charset.StandardCharsets.US_ASCII;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertNotEquals;
+import static org.junit.jupiter.api.Assertions.assertNotNull;
+
+class HttpResponseConnectionLabelTest {
+
+    private static final Logger LOGGER =
+            Utils.getDebugLogger(HttpResponseConnectionLabelTest.class.getSimpleName()::toString, Utils.DEBUG);
+
+    private static final Charset CHARSET = US_ASCII;
+
+    private static final String CONNECTION_KEY_HEADER_NAME = "X-Connection-Key";
+
+    private static final String SERVER_ID_HEADER_NAME = "X-Server-Id";
+
+    private static final SSLContext SSL_CONTEXT = createSslContext();
+
+    // Start with a fresh client having no connections in the pool
+    private final HttpClient client = HttpClient.newBuilder().sslContext(SSL_CONTEXT).proxy(NO_PROXY).build();
+
+    // Primary server-client pairs
+
+    private static final ServerRequestPair PRI_HTTP1 = ServerRequestPair.of(Version.HTTP_1_1, false);
+
+    private static final ServerRequestPair PRI_HTTPS1 = ServerRequestPair.of(Version.HTTP_1_1, true);
+
+    private static final ServerRequestPair PRI_HTTP2 = ServerRequestPair.of(Version.HTTP_2, false);
+
+    private static final ServerRequestPair PRI_HTTPS2 = ServerRequestPair.of(Version.HTTP_2, true);
+
+    // Secondary server-client pairs
+
+    private static final ServerRequestPair SEC_HTTP1 = ServerRequestPair.of(Version.HTTP_1_1, false);
+
+    private static final ServerRequestPair SEC_HTTPS1 = ServerRequestPair.of(Version.HTTP_1_1, true);
+
+    private static final ServerRequestPair SEC_HTTP2 = ServerRequestPair.of(Version.HTTP_2, false);
+
+    private static final ServerRequestPair SEC_HTTPS2 = ServerRequestPair.of(Version.HTTP_2, true);
+
+    private static SSLContext createSslContext() {
+        try {
+            return new SimpleSSLContext().get();
+        } catch (IOException exception) {
+            throw new UncheckedIOException(exception);
+        }
+    }
+
+    private record ServerRequestPair(
+            HttpTestServer server,
+            HttpRequest request,
+            boolean secure,
+            CountDownLatch[] serverResponseLatchRef) {
+
+        private static final AtomicInteger SERVER_COUNTER = new AtomicInteger();
+
+        private static final AtomicInteger SERVER_RESPONSE_COUNTER = new AtomicInteger();
+
+        private ServerRequestPair {
+            try {
+                server.start();
+            } catch (Exception serverException) {
+                throw new RuntimeException("failed closing server", serverException);
+            }
+        }
```

Comment:
The message here seems to be wrong. Can we use something less exotic to start the server too, and start it, for innstance in the calling code?

Reply:
And also maybe log something after it's started - including the server authority where it's listening at?

Reply:
Both are implemented in a0859c7c68614e8c3f531b1d4d9357f3f48cf5ab.